include/odp/api/spec/classification.h

```diff
@@ -164,6 +171,21 @@ typedef struct odp_bp_param_t {
 	odp_threshold_t threshold;
 } odp_bp_param_t;
 
+/** Mark support
+ * When mark support is enabled for a particular flow, the implementation shall
```

Comment:
> When mark support is enabled for a particular flow,

That may be slightly confusing since there is no separate enable control but just the desired mark value. I would also change "insert" to "set" because there cannot be more than one CoS mark set in packet metadata.

The default value set by odp_cls_cos_param_init() should be defined.

I would rephrase this somehow like this:

CoS mark support

When a packet is processed by a class-of-service, CoS mark in packet metadata is set to the value configured for the class-of-service through odp_cls_cos_param_t.

Reply:
> > When mark support is enabled for a particular flow,
> 
> That may be slightly confusing since there is no separate enable control but just the desired mark value. I would also change "insert" to "set" because there cannot be more than one CoS mark set in packet metadata.

Yes. "set" makes sense.

> 
> The default value set by odp_cls_cos_param_init() should be defined.

Ack.

> 
> I would rephrase this somehow like this:
> 
> CoS mark support
> 
> When a packet is processed by a class-of-service, CoS mark in packet metadata is set to the value configured for the class-of-service through odp_cls_cos_param_t.

After implementing this patch on one of the Marvell platform, I realized that, Hooking Mark in CoS would waste the HW resources like RQ in the case where "single" CoS preferred for application and just because of  CoS has match value, application needs to create another CoS and it have bearing on HW resources that way it is mapped to Implementation. Look like the ideal mapping is NOT CoS for mark, Instead have new variant of odp_cls_pmr_create() as odp_cls_pmr_create_with_mark().

/**
 * Create a packet matching rule with mark id
 *
 * It is the mark feature variant of odp_cls_pmr_create().
 * When a PMR is created with mark support, the implementation shall insert
 * the specified mark_id in the packet metadata. If the packet matches the PMR
 * then the application can retrieve it through odp_packet_mark_id().
 *
 * @param terms        Array of odp_pmr_param_t entries, one entry per term
 *                     desired.
 * @param num_terms    Number of terms in the match rule.
 * @param mark_id      mark id to insert in the packet metadata on match.
 * @param src_cos      source CoS handle
 * @param dst_cos      destination CoS handle
 *
 * @return Handle to the Packet Match Rule.
 * @retval ODP_PMR_INVALID on failure
 * @note The mark id value is arbitrary and application-defined.
 * The maximum allowed value depends on the underlying implementation.
 * @see odp_cls_capability_t::max_mark_id for maximum mark id value
 * @see odp_packet_mark_id() to get the mark id for the packet.
 */
odp_pmr_t odp_cls_pmr_create_with_mark(const odp_pmr_param_t *terms, int num_terms,
                                       uint64_t mark_id, odp_cos_t src_cos, odp_cos_t dst_cos);

Reply:
I am not yet sure, but I think I am in favor of putting mark somehow in the PMR (but with a cleaner and more generic API that the odp_cls_pmr_create_with_mark()).

Initially I thought that you could just map the abstract API level PMR + CoS configuration to somewhat different HW level "PMR" + "CoS" configuration. For instance, if an application would like to set mark based on IP address and select destination queue based on DSCP, the application would just create all the different CoS combinations, but the implementation could notice when a new CoS differs from an existing one only in mark and then reuse the existing CoS and do the marking as part of the different packet patching rules that now point to the shared "CoS" in the HW.

This way PMR rules would remain as pure matchers and all actions would be defined in the CoS objects.

But perhaps that is not the right model. Both Linux netfilter and DPDK flow API allow mark (and other) actions as part of the matching rule. We could do the same in ODP, but that would be a significant conceptual change that needs to be discussed with Petri too.

Btw, there is an additional problem in putting mark in CoS and having application create all the combinations even if the implementation could optimize the redundancy away: When using multiple queues with hashing in a CoS, the implementation would have to create a separate set of queues for each QoS. Thus an application can create multiple CoS which differ only in mark only if it can specify the exact same queue(s) for the CoS and that is now possible only with single queue per CoS.

Reply:
I discussed with Janne about this, and it seems that adding mark to pmr is the best solution. Unfortunately, pmr create does not have struct that we could modify, so a new create function seems the only option. For future proof design, it would be good to introduce a structure this time. So, we'd suggest something like:

typedef struct odp_pmr_create_opt_t {
    const odp_pmr_param_t *terms;
    int num_terms;
    uint64_t mark_id;
    odp_cos_t src_cos;
    odp_cos_t dst_cos;
} odp_pmr_create_opt_t;

odp_pmr_t odp_cls_pmr_create_opt(const odp_pmr_create_opt_t *opt);

src_cos and dst_cos could be also direct function parameters. That would keep current look/feel of the function, where the call connects src_cos to dst_cos. 

odp_pmr_t odp_cls_pmr_create_opt(const odp_pmr_create_opt_t *opt, odp_cos_t src_cos, odp_cos_t dst_cos);

Also capability is needed for the maximum mark size (0 ... max_mark_id).

Reply:
Ack. How about odp_pmr_create_meta_t instead of odp_pmr_create_opt_t and odp_cls_pmr_create_with_meta() to denote this variant of function would be useful to create the PMR with metadata(the content which not available in the packet unlike normal terms like match id and future some other extension like IPsec second pass metadata from HW accelerators etc). Thoughts on the name?

Reply:
OR
typedef struct odp_pmr_apply_meta_t {
             uint64_t mark_id;
} odp_pmr_apply_meta_t;
odp_cls_pmr_apply_meta(odp_pmr_t pmr, odp_pmr_apply_meta_t meta);
This would apply meta based patterns to exisitng odp_pmr_t  the one created with odp_cls_pmr_create().

Reply:
I would prefer opt over meta as it is maybe more generic. Maybe there will later be need to add something that is not metadata specific at all. We even thought about odp_cls_pmr_create2().

Postponing the configuration from PMR creation to a later attach/apply step is problematic because it would require the implementation be able to modify an existing PMR on the fly when it is already being used for packet matching. The implementation cannot postpone the creation until the attach/apply phase because it does not know if it ever comes.

Reply:
OK. Makes sense. In order to add comments, I would like to know the full form of "opt" , Is it "option"?
Is there any name conversion on the table for the new variant? Is extension -> odp_cls_pmr_create_extn(), alternative -> odp_cls_pmr_create_alt(), varint -> odp_cls_pmr_create_varint() can be used instead of option. No strong option on the name, Waiting for the feedback on the name for v2.

Reply:
I think Petri should comment.

I am not a big fan of "opt", but there are already various _opt_t types in ODP for miscellaneous config options so it seems to me that opt would be in line with those. "alt", "variant", "ext" and such would be totally new. That said, I do not have a strong opinion on naming.

Reply:
I will stick with odp_cls_pmr_create_opt() then.

Reply:
Sent the  V2 with _opt scheme. Resolving this one.